### ISSUE TITLE
Introduction of a Stability Term for Numerical Stability.

### ISSUE THREAD

**Issue by jon-tow** [tensorflow/swift-apis#237]
Are there any plans to introduce a stability term such as the [epsilon constant in the Keras backend](https://www.tensorflow.org/api_docs/python/tf/keras/backend/epsilon)? This could be useful in the Loss api for losses that depend on ratios.

**Comment by eaplatanios** [comment]
Swift already has [an API for that for floating point numbers](https://developer.apple.com/documentation/swift/floatingpoint/3017999-ulpofone).

**Comment by jon-tow** [comment]
For some reason I assumed this wasn't built into the language, sorry about that.

**Comment by jon-tow** [comment]
@eaplatanios For some optimizer classes, there are these `epsilon` instance properties used to improve numerical stability. Would you suggest replacing these fields with `ulpOneOf`?

**Comment by eaplatanios** [comment]
No worries. We could set the optimizer `epsilon`s default values to `ulpOfOne`, but I would suggest keeping the arguments so that users can control the amount of smoothing applied during optimization. Numerical optimization is often unstable and sensitive to smoothing so it's good being able to control for that.

**Comment by jon-tow** [comment]
Good idea. Thanks!

**Comment by dan-zheng** [comment]
The API documentation for [`ulpOfOne`](https://developer.apple.com/documentation/swift/floatingpoint/3017999-ulpofone) seems to suggest that it's not suitable for approximate equality?

> Swift deliberately avoids using the term “epsilon” because:
> * The name “epsilon” suggests that this quantity is a good tolerance to choose for approximate comparisons, but it is almost always unsuitable for that purpose.

There was a recent pitch to add [approximate floating-point equality](https://forums.swift.org/t/approximate-equality-for-floating-point/22420?u=dan-zheng) to the standard library, which uses `sqrt(.ulpOfOne)` as the default tolerance value.

```swift
  /// The default value of `tolerance` is `sqrt(.ulpOfOne)`; this value
  /// comes from the common numerical analysis wisdom that if you don't
  /// know anything about a computation, you should assume that roughly
  /// half the bits may have been lost to rounding. This is generally a
  /// pretty safe choice of tolerance--if two values that agree to half
  /// their bits but are not meaningfully almost equal, the computation
  /// is likely ill-conditioned and should be reformulated.
  ///
  /// ...
  public func isAlmostEqual(
    to other: Self,
    tolerance: Self = Self.ulpOfOne.squareRoot()
  ) -> Bool
```
